### PR TITLE
RavenDB-18778: Fixes corruptions at the CompactTree level

### PR DESCRIPTION
**Pull request by redknightlois**
This PR is deemed critical and fixes 4 different corruption issues happening at the tree splitting/merging process. There is still one open minor issue with calculation of the depth metric that is deemed non-critical, as will be addressed by: https://issues.hibernatingrhinos.com/issue/RavenDB-18801

EDIT: Before it said that some tests would fail. That is no longer the case because the error only manifest under certain conditions and we need to use the Tryout to repro the parameters. As is they work.

### CODE REVIEW

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 1]
```diff
@@ -490,16 +491,19 @@ private bool RemoveFromPage(bool allowRecurse, out long oldValue)
...
+
                 MaybeMergeEntries(ref state);
             }
 
             return true;
         }
-
-        private void Verify()
+        
+        private void Verify(ref CursorState current)
```
Comment (ayende) [inline src/Voron/Data/CompactTrees/CompactTree.cs:504]: That should be `[Conditional]` ?

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 2]
```diff
@@ -532,15 +538,15 @@ private void Verify()
...
+                    VoronUnrecoverableErrorException.Raise(_llt, "Last encoded key does not follow lexicographically.");
                 }
                 
                 lastEncodedKey = encodedKey;
                 lastDecodedKey = decodedKey;
             }
         }
 
-        private void MaybeMergeEntries(ref CursorState destinationState)
+        private bool MaybeMergeEntries(ref CursorState destinationState)
```
Comment (ayende) [inline src/Voron/Data/CompactTrees/CompactTree.cs]: Who is using the return value?

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 3]
```diff
@@ -768,7 +778,7 @@ private EncodedKey AddToPage(EncodedKey key, long value)
                 if (len == valueEncoder.Length)
                 {
                     Debug.Assert(valueEncoder.Length <= sizeof(long));
-                    Memory.Copy(b, valueEncoder.Buffer, valueEncoder.Length);
+                    Unsafe.CopyBlockUnaligned(b, valueEncoder.Buffer, (uint)valueEncoder.Length);
```
Comment (ayende) [inline src/Voron/Data/CompactTrees/CompactTree.cs:775]: Is that better?

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 4]
```diff
@@ -1080,18 +1092,16 @@ private void InsertToStack(CursorState newPageState)
...
+        private static void DefragPage(LowLevelTransaction llt, ref CursorState state)
         {                     
-            ref var state = ref _internalCursor._stk[_internalCursor._pos];
-
-            using (_llt.Environment.GetTemporaryPage(_llt, out var tmp))
+            using (llt.Environment.GetTemporaryPage(llt, out var tmp))
             {
-                var tmpHeader = (CompactPageHeader*)tmp.TempPagePointer;
+                // Ensure we clean up the page.               
+                Unsafe.InitBlock(tmp.TempPagePointer, 0, Constants.Storage.PageSize);
```
Comment (ayende) [inline src/Voron/Data/CompactTrees/CompactTree.cs:1093]: That is faster?

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 2 above)
Comment (redknightlois) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: This was useful for debugging. Should I kill it?

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 3 above)
Comment (redknightlois) [reply src/Voron/Data/CompactTrees/CompactTree.cs:775]: We need to eventually revisit memory copy. For small copies there is nothing faster than `.CopyBlockUnaligned()` and most likely for larger too. It has been heavily optimized with SIMD instructions and pretty fine microtuning.

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 4 above)
Comment (redknightlois) [reply src/Voron/Data/CompactTrees/CompactTree.cs:1093]: Same as `.CopyBlockUnaligned()` and `.CopyBlock()`

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 1 above)
Comment (redknightlois) [reply src/Voron/Data/CompactTrees/CompactTree.cs:504]: No, most likely if you need to use it to verify you dont want to run in DEBUG

Comment (ayende) [reply src/Voron/Data/CompactTrees/CompactTree.cs:504]: Who is calling this? That can be expensive

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 2 above)
Comment (ayende) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: I think so

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 1 above)
Comment (redknightlois) [reply src/Voron/Data/CompactTrees/CompactTree.cs:504]: Nobody. It is a debugging resource.

File: test/FastTests/Voron/CompactTreeTests.cs [hunk 5]
```diff
@@ -185,7 +185,55 @@ public void CanDeleteLargeNumberOfItemsInRandomOrder()
                 Assert.Equal(0, tree.State.NumberOfEntries);
                 Assert.Equal(0, tree.State.BranchPages);
                 Assert.Equal(1, tree.State.LeafPages);
-                Assert.Equal(1, tree.State.Depth);
+            }
+        }
+
+        [RavenTheory(RavenTestCategory.Voron)]
+        [InlineData(400000)]
```
Comment (arekpalinski) [inline test/FastTests/Voron/CompactTreeTests.cs]: @redknightlois I see we have more tests playing with 400000 entries. I will move those cases to SlowTests

Comment (redknightlois) [reply test/FastTests/Voron/CompactTreeTests.cs]: No problem

**Pull request merged**